chalk-solve/src/clauses.rs

```diff
@@ -222,19 +233,24 @@ fn program_clauses_that_could_match<I: Interner>(
                 // and `bounded_ty` is the `exists<T> { .. }`
                 // clauses shown above.
 
-                for exists_qwc in dyn_ty.bounds.map_ref(|r| r.iter(interner)) {
-                    // Replace the `T` from `exists<T> { .. }` with `self_ty`,
-                    // yielding clases like
-                    //
-                    // ```
-                    // forall<'a> { Implemented(dyn Fn(&u8): Fn<(&'a u8)>) }
-                    // ```
-                    let qwc = exists_qwc.substitute(interner, &[self_ty.clone().cast(interner)]);
-
-                    builder.push_binders(&qwc, |builder, wc| {
-                        builder.push_fact(wc);
-                    });
-                }
+                let generalized_dyn_ty = generalize::Generalize::apply(db.interner(), dyn_ty);
```

Comment:
What is this...? Needs a comment at least

Reply:
OK, read the comments on the generalize module, I think I see what's going on here. I do think we should comment somewhere. If I understand, you might have a type like

```rust
dyn Foo<?X>
```

and we will "generalize" this to `exists<A> { dyn Foo<A> }` and then make a rule like

```
forall<A> {
    Implemented(dyn Foo<A>: Foo<A>)
}
```

which... does seem right.

Reply:
Exactly -- it's also kind of related to the fact that the recursive solver calls `program_clauses` before instantiation, the SLG solver would probably have inference variables there and not have a problem :thinking: